Bluetooth: host: Fix GATT clear bond information bugs

joerchan opened the pull request:
Fix CCC cfg not cleared when overwriting oldest bond. Calling
bt_unpair with a pointer to the key will result in the key addr being
memset to zero and bt_gatt_clear is called with an zero-set address.
This happens because unpair (hci_core.c) calls bt_keys_clear before
calling bt_gatt_clear.

Fixes: #23961

Fix Service Changed configuration stored in flash not deleted in flash
when calling bt_gatt_clear from bt_unpair.
When clearing the check for "is bonded" should not be made. If the
bond information is already removed this check will fail.
When clearing the check for "modified" should not be made, clearing
is in itself a modification, and should always be made.

joerchan committed:
Bluetooth: host: Fix CCC cfg not cleared when overwriting oldest bond

Fix CCC cfg not cleared when overwriting oldest bond. Calling
bt_unpair with a pointer to the key will result in the key addr being
memset to zero and bt_gatt_clear is called with an zero-set address.
This happens because unpair (hci_core.c) calls bt_keys_clear before
calling bt_gatt_clear.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan commented:
CC: @mfiumara

joerchan commented:
@jhedberg This could have been fixed by re-ordering bt_gatt_clear and bt_keys_clear. I'm wondering if we should do both fixes. Assuming re-ordering has no negative effects.

jhedberg commented:
@joerchan `bt_gatt_clear()` ends up calling `bt_addr_le_is_bonded()` when clearing Service Changed information, so I think the order must be kept (would be good to have a code comment though so no-one goes and changes this)

Vudentz commented:
> @joerchan `bt_gatt_clear()` ends up calling `bt_addr_le_is_bonded()` when clearing Service Changed information, so I think the order must be kept (would be good to have a code comment though so no-one goes and changes this)

That is just removing the entry from the settings, not sure if that is already done implicitly by bt_unpair, it doesn't looks like it does anything related to settings though, it seems it just clear the keys entry so actually doing the bt_gatt_clear before bt_keys_clear should make sure the settings storage is cleanup properly.

joerchan commented:
@Vudentz It looks like the SC would not be cleared from flash then. The modified check would also affect this.

jhedberg commented:
@Vudentz @joerchan so what's the consensus on this PR? Do we merge it it it needs additional changes?

joerchan commented:
> @Vudentz @joerchan so what's the consensus on this PR? Do we merge it it it needs additional changes?

I would say merge this one. And I'll look if there is a separate problem and submit a PR if that is the case.

joerchan committed:
Bluetooth: host: Fix Service Changed not cleared from flash

Fix Service Changed configuration stored in flash not deleted in flash
when calling bt_gatt_clear from bt_unpair.
When clearing the check for "is bonded" should not be made. If the
bond information is already removed this check will fail.
When clearing the check for "modified" should not be made, clearing
is in itself a modification, and should always be made.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

jhedberg commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -252,44 +252,68 @@ static void sc_store(struct gatt_sc_cfg *cfg)
-			}
+static void sc_clear(struct bt_conn *conn)
+{
+	if (!bt_addr_le_is_bonded(conn->id, &conn->le.dst)) {
```
Could you flip the main- and else-branches so that you don't have to do the negation for the condition?

Vudentz reviewed (approved):
LGTM
